Python 3 Port[In progress]

Saumya-Mishra9129 opened the pull request:
Trying to port from python 2 to 3 , as Sugar-toolkit is also ported , I am getting this error, get stuck at this.Kindly look at error.
![Screenshot from 2020-03-25 00-18-15](https://user-images.githubusercontent.com/45252461/77465170-3ed63c80-6e2e-11ea-9bb9-dab0ca3f4720.png)

and do review also.Thanks

Saumya-Mishra9129 committed:
Port-to-Python3,Gtk3.0&GLib

Saumya-Mishra9129 committed:
deleted vte and libvte

chimosky commented:
The activity hasn't been ported to Gtk3, you should do that first. The error you got is as a result of that.
Gtk3 activities use `toolbarbox` and not `toolbox`.

quozl commented:
@Saumya-Mishra9129, next time please copy and paste errors into GItHub markdown text blocks, using _```_ to delimit.  Images are expensive, and can't be searched.

Saumya-Mishra9129 commented:
@quozl  Okay sure. 
@chimosky  thanks. Port to Gtk3 is still in progress. I am working on it.

chimosky commented:
> Port to Gtk3 is in progress. I am working on it.

Great, we'll wait for it.

Saumya-Mishra9129 committed:
fixed typo

Saumya-Mishra9129 committed:
Fixes toolbar

Saumya-Mishra9129 committed:
fix Vte errors

Saumya-Mishra9129 committed:
fix Gdk.color_parse error

Saumya-Mishra9129 committed:
fix minor typo in edit_toolbar

Saumya-Mishra9129 commented:
@pro-panda @quozl I have tried to fix toolbar and Vte errors . Please Review. Also some of the methods are removed from Vte.Terminal class in Vte 2.91 as fork_command(), set_emulation() and set_visible_bell(). I couldn't figure out what to do in such cases. Can you suggest here.??

chimosky commented:
I think `spawn_async` replaced `fork_command`, the [changelog](https://gitlab.gnome.org/GNOME/vte/-/blob/master/ChangeLog.pre-git#L5850) confirms that, one thing you can do is check what those functions do in python2 by importing `vte` in a python2 REPL and find out how to do those things with `Vte.Terminal`.

Saumya-Mishra9129 commented:
@srevinsaju I looked at Terminal activity which also contains Vte. You have worked there , Can you have a look at this activity?

srevinsaju commented:
@Saumya-Mishra9129 i will check

quozl commented:
@Saumya-Mishra9129, the Vte library was changed a few times, may I suggest;

* search for the activities that instantiate `Vte.Terminal`,
* look through the commits in those activities for changes in support of the various versions of Vte,
* search older versions of those activities for those removed methods, then search for commits that ported to later Vte.

srevinsaju committed:
Minor formatting

srevinsaju committed:
Reorder and optimize imports

srevinsaju committed:
Add extra line before class definition PEP8

srevinsaju committed:
Remove trailing spaces; add newlines wherever necessary

srevinsaju committed:
Use str as arg in configparser

configparser accepts <class 'str'> as *args.
Boolean and integer args had to be converted to str

srevinsaju committed:
Use single quote instead of double to match i18n conventions

srevinsaju committed:
Convert string config args to parseable set_cursor_blink_mode

Fixes TypeError: Expected a Vte.CursorBlinkMode, but got str

srevinsaju committed:
fix TypeError on set_scrollback; Fixes r/w on config

TypeError: Must be number, not str

srevinsaju committed:
Conditionally set Vte.set_emulation if Vte is of an older commit

Since 4e253be9282829f594c8a55ca08d1299e80e471d on Vte's codebase
Vte has removed the set_emulation attribute
Only set it if necessary

srevinsaju committed:
Conditionally set_visible_bell if Vte.Terminal is older commit

Since 4e253be9282829f594c8a55ca08d1299e80e471d on Vte's codebase
Vte has removed the set_visible_bell attribute
Only set it if necessary

srevinsaju committed:
Minor white space fixes

srevinsaju committed:
Rewrite feed_child to match fix line_length

srevinsaju committed:
Port fork_command to spawn_sync

srevinsaju commented:
@Saumya-Mishra9129 Please review https://github.com/Saumya-Mishra9129/Frotz/pull/1

Currently the `frotz` binary doesn't work;
This is possibly because some of the needed binaries doesn't work
Please recompile frotz and include all the necessary libs within `lib` to make this work

To include `LD_LIBRARY_PATH`, you can mention it in `envv` args in `spawn_sync`. I have not used `spawn_async`, instead used `spawn_sync`. I hope it would be equally good

Saumya-Mishra9129 committed:
update frotz binary file : FROTZ V2.44

Saumya-Mishra9129 commented:
@srevinsaju I just looked at VTE documentation and found that spawn_sync is deprecated since version 0.48. For your reference -https://lazka.github.io/pgi-docs/Vte-2.91/classes/Terminal.html#Vte.Terminal.spawn_sync

srevinsaju committed:
Fix deprecation of spawn_sync. Port spawn_sync -> spawn_async

srevinsaju commented:
@Saumya-Mishra9129 
See https://github.com/Saumya-Mishra9129/Frotz/pull/2

-----------

Although deprecated; its still usable; Anyways, I have used `spawn_async` on PR 2

chimosky commented:
Reviewed f38db26.

An issue wasn't opened for frotz binaries in this activity but I think you should remove the binaries, the binaries exists for users who would run the activity on an XO laptop; porting to python3 removes the need for that.
- [ ] Change to using system specific package for frotz and specify dependency in README.

quozl commented:
> * [ ]  Change to using system specific package for frotz and specify dependency in README.

Agreed.  This should also result in a fix for https://github.com/sugarlabs/Frotz/issues/2.

Reminder of standing advice for future use: [_make sure the activity works properly before porting, close any solved issues, merge any pull requests or branches and release the last Python 2 version_](https://github.com/sugarlabs/sugar-docs/blob/master/src/python-porting-guide.md#how-to-port-an-activity-to-python-3).

quozl commented:
Re: https://github.com/sugarlabs/Frotz/pull/7/commits/d57cfd50f6c1f0b50f225e1dc9a70982045a1a19, my preference is to use `Vte.Terminal.spawn_sync` until it is unavailable.  `Vte.Terminal.spawn_async` is new, and not available, for example, in Ubuntu 18.04.  Moving quickly just because someone shouts **deprecated** at us breaks compatibility with older operating systems without a good reason.

Or, you can use `hasattr` and support both methods.

Saumya-Mishra9129 commented:
> Re: d57cfd5, my preference is to use Vte.Terminal.spawn_sync until it is unavailable. Vte.Terminal.spawn_async is new, and not available, for example, in Ubuntu 18.04. Moving quickly just because someone shouts deprecated at us breaks compatibility with older operating systems without a good reason.

Thanks , I tested spawn_async is not working with Ubuntu 18.04.

> Or, you can use hasattr and support both methods

Yeah It can be best.

Saumya-Mishra9129 committed:
Add compatibility of spawn_async and spawn_sync

Saumya-Mishra9129 committed:
Optimize Imports

Saumya-Mishra9129 committed:
Remove binary files and use system specific package for frotz

Saumya-Mishra9129 committed:
Specify dependency for frotz in README

Saumya-Mishra9129 commented:
> Change to using system specific package for frotz and specify dependency in README.

I have tried this in b3bc77d and c52c1da. Activity still doesn't work , maybe something is broken.

Saumya-Mishra9129 committed:
Fix Vte.Terminal

Terminal wasn't loading, crashing with no errors in log

Saumya-Mishra9129 committed:
Fix Typo in b3bc77d

Saumya-Mishra9129 commented:
> I have tried this in b3bc77d and c52c1da. Activity still doesn't work and with no errors in the log , maybe something is broken.

Finally get it working fine. :smile:  @quozl @chimosky @srevinsaju @JuiP Please review.

quozl commented on README.md:
```diff
@@ -0,0 +1,31 @@
+
+Frotz.activity is a simple front end to the classic interactive fiction interpreter frotz on the XO laptop.  Frotz.activity is licensed under the GNU General Public License version 3 (see the file COPYING for details). 
+
+The actual interactive interpreter, Frotz, is licensed under the GNU General Public License version 2.  This xo bundle includes only a binary (version 2.43), but you can obtain the source from here: http://frotz.homeunix.org/frotz/
```
Says here it only includes a binary, but you removed the binary.

Saumya-Mishra9129 replied:
Yeah Done

quozl commented on README.md:
```diff
@@ -0,0 +1,31 @@
+What is this?
+=============
+
+Frotz.activity is a simple front end to the classic interactive fiction interpreter frotz on the XO laptop.  Frotz.activity is licensed under the GNU General Public License version 3 (see the file COPYING for details). 
```
Let's clearly differentiate between the activity and the upstream software it is a wrapper for.  I suggest changing "Frotz.activity" to "Frotz activity" in this paragraph, and from line 8 onwards when the context is a reference to the activity instead of the Frotz upstream software.

Saumya-Mishra9129 replied:
Thanks changed!

quozl commented on README.md:
```diff
@@ -0,0 +1,31 @@
+How to install dependencies?
+===========================================
+
+Install the frotz
```
"the frotz"?  :grin:

Saumya-Mishra9129 replied:
:sweat_smile:

quozl commented on activity/activity.info:
```diff
@@ -2,8 +2,8 @@
 license = GPLv3+
-host_version = 1
\ No newline at end of file
+host_version = 1
```
host version is no longer required.

Saumya-Mishra9129 replied:
Removed

quozl commented on frotz.py:
```diff
@@ -126,15 +131,34 @@ def start_game(self, game_file):
+                    text="frotz is not installed"
+                )
+                dialog.format_secondary_text(
+                    "Please Install frotz , follow README"
```
A space before the comma.  Also the file README is not easily accessible to the child user, so this doesn't help.

Saumya-Mishra9129 replied:
Thanks @quozl , will it work this way c388c5c ?

quozl commented on frotz.py:
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
             scroll_key = conf.getboolean('terminal', 'scroll_on_keystroke')
         else:
-            scroll_key = False
+            scroll_key = 'False'
```
Should be a boolean.  Reference the Terminal activity and PyGObject docs for the function.

Saumya-Mishra9129 replied:
Because , configparser accepts <class 'str'> as *args.

Saumya-Mishra9129 replied:
@quozl See https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.set, I wonder about terminal activity, do we need changes to terminal activity.?

quozl replied:
I've seen no such problem on Terminal.

quozl commented on frotz.py:
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
             scroll_output = conf.getboolean('terminal', 'scroll_on_output')
         else:
-            scroll_output = False
+            scroll_output = 'False'
```
Ditto.

quozl commented on frotz.py:
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
             bell = conf.getboolean('terminal', 'bell')
         else:
-            bell = False
+            bell = 'False'
```
Should be a boolean.

Saumya-Mishra9129 replied:
Because` configparser accepts <class 'str'> as *args.`

quozl replied:
Not a good enough reason.  What matters here is what Vte requires.

Saumya-Mishra9129 replied:
https://github.com/sugarlabs/Frotz/pull/7/commits/813abae39151c64aebca002642f97413fc204c07 Thanks Please review.

quozl commented on frotz.py:
```diff
@@ -235,61 +283,63 @@ def _configure_vte(self):
-            
-        self.set_scrollback_lines(scrollback_lines)
+
+        self.set_scrollback_lines(int(scrollback_lines))
```
Not sure why this changes.

Saumya-Mishra9129 replied:
Because of `TypeError: Must be number, not str`, scrollback_lines takes int as argument , not str.

quozl replied:
Change line 319 back to integer, not string.

quozl replied:
Not a good enough reason to change, in my opinion.

Saumya-Mishra9129 replied:
@quozl https://github.com/sugarlabs/Frotz/pull/7/commits/813abae39151c64aebca002642f97413fc204c07 This should be fine.

srevinsaju commented on frotz.py:
```diff
@@ -13,94 +13,99 @@
         scrollbar.show()
 
-        box.pack_start(self._vte)
+        box.pack_start(self._vte, True, True, 0)
```
providing `kwargs` might make the code more readable

Saumya-Mishra9129 replied:
Done

srevinsaju commented on frotz.py:
```diff
@@ -126,15 +131,34 @@ def start_game(self, game_file):
-            if platform.machine().startswith('arm'):
-                logging.error('ARM not supported, yet') # FIXME
-                sys.exit(0)
+            cmd = '/usr/games/frotz'
```
Are you sure `frotz` is always installed in `/usr/games/frotz` across distros?

Saumya-Mishra9129 replied:
Yeah , I have checked on Debian and Ubuntu, Not sure about fedora. Do you use Fedora, Can you check once?

srevinsaju replied:
Fedora saves it to the /usr/bin as other distributions do. Only Linux distros that use `apt` saves it to `/usr/games`. All others, Fedora, RHEL, CentOS, Arch Linux and Manjaro installs it to `/usr/bin`

Saumya-Mishra9129 replied:
Thanks Done in 5c38b6c

quozl replied:
/usr/games is in $PATH on Debian and Ubuntu, so use $PATH.

srevinsaju commented on frotz.py:
```diff
@@ -143,19 +167,19 @@ def read_file(self, file_path):
             tmpfile.close()
-            os.chmod(tmpfolder, 0755)
-            os.chmod(tmpfilepath, 0755)
+            os.chmod(tmpfolder, 0o755)
+            os.chmod(tmpfilepath, 0o755)
```
out of curiosity and due to security reasons, maybe we should not make a temp file executable for the world?

Saumya-Mishra9129 replied:
Yeah I was also curious to know why they are doing this. Isn't there any other way? :thinking:

srevinsaju replied:
Change it to 644 perhaps

Saumya-Mishra9129 replied:
Great , done in 607dd6f

srevinsaju commented on frotz.py:
```diff
@@ -187,30 +211,54 @@ def _paste_cb(self, button):
+        Vte.Terminal.__init__(self)
+        self.configure_terminal()
+
+        os.chdir(os.environ["HOME"])
```
can you provide an inline comment explaining why this was uncommented or why this is necessary?

Saumya-Mishra9129 replied:
Thanks for noticing, The line seems to have no effect. Actually activity was crashing before, so I copied some of the code from VncLauncher activity. I think it is good to remove line, if it has no effect.

srevinsaju replied:
Great

srevinsaju commented on setup.py:
```diff
@@ -1,4 +1,4 @@
 #!/usr/bin/python
```
It would be better to use `/usr/bin/env python3`, as `python` does not exist on Ubuntu 20.04

Saumya-Mishra9129 replied:
Done!

Saumya-Mishra9129 committed:
Update README.md

Saumya-Mishra9129 committed:
Remove host_version

Saumya-Mishra9129 committed:
Change shebang to Python3

Saumya-Mishra9129 committed:
Add kwargs

Saumya-Mishra9129 committed:
Revert : Change Uncommented to commented line

Saumya-Mishra9129 committed:
Modify message dialog to install frotz

Saumya-Mishra9129 committed:
Modify command to start the game

Fedora, RHEL, CentOS, Arch Linux and Manjaro installs frotz to /usr/bin. Only Linux distros that use apt saves it to /usr/games.

Saumya-Mishra9129 committed:
Change temp file permission to 0o644

0o755 gives execute permission to temp file. Because of security reasons, we should not make a temp file executable for the world. Change to 0o644 gives only read-write permission only.

chimosky commented on frotz.py:
```diff
@@ -131,37 +131,34 @@ def start_game(self, game_file):
+                                save_dir,
+                                os.path.basename(game_file),
+                                game_file)).encode('utf-8')
+                )
```
Why this indentation?

Saumya-Mishra9129 replied:
Done in 480c207.

quozl commented on README.md:
```diff
@@ -0,0 +1,31 @@
+What is this?
+=============
+
+Frotz activity is a simple front end to the classic interactive fiction interpreter Frotz on the XO laptop.  Frotz activity is licensed under the GNU General Public License version 3 (see the file COPYING for details). 
```
No longer for the XO laptop.

quozl replied:
Details of the licensing are actually in frotz.py, not in COPYING.  COPYING does contain details, but not the copyright and license.  If in doubt as to the best thing to say; what is in frotz.py header is exactly what should be said.

Saumya-Mishra9129 replied:
Done in c89585f

quozl commented on README.md:
```diff
@@ -0,0 +1,31 @@
+
+Frotz activity is a simple front end to the classic interactive fiction interpreter Frotz on the XO laptop.  Frotz activity is licensed under the GNU General Public License version 3 (see the file COPYING for details). 
+
+The actual interactive interpreter, Frotz, is licensed under the GNU General Public License version 2. You can obtain the source from here: http://frotz.homeunix.org/frotz/
```
Link does not work.

quozl replied:
License of Frotz includes the _or later version_ wording.

Saumya-Mishra9129 replied:
Done in c89585f

quozl commented on README.md:
```diff
@@ -0,0 +1,31 @@
+
+
+How to install dependencies?
+===========================================
```
Underline extends beyond heading.

Saumya-Mishra9129 replied:
Done in c89585f

quozl commented on frotz.py:
```diff
@@ -112,10 +117,10 @@ def __init__(self, handle):
-        print "Quitting..."
+    def _quit_cb(self, vte, foo=None):
+        logging.debug("Quitting...")
         sys.exit(0)
```
Use Activity.close instead?  See what Terminal does; when last tab is clsoed it calls `self.close()`.

Saumya-Mishra9129 replied:
Done in 298e911

quozl commented on frotz.py:
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
-            if platform.machine().startswith('arm'):
-                logging.error('ARM not supported, yet') # FIXME
-                sys.exit(0)
+            if platform.version().find("Ubuntu") > -1 or platform.version().find("Debian") > -1:
```
While this works for Ubuntu and Debian, a more correct method is to search for the binary in a list of paths, like how Linux does it.

quozl commented on frotz.py:
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
+            if os.path.isfile(cmd) and os.access(cmd, os.X_OK):
+                logging.debug('Using frotz installed in the system')
+                self._vte.feed_child(
+                    (
```
Indentation here is very strange; the open parenthesis is not aligned with the close.

Saumya-Mishra9129 replied:
Done in 480c207

quozl commented on frotz.py:
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
+                    text="frotz is not installed"
+                )
+                dialog.format_secondary_text(
+                    "Install frotz by clicking OK and then restart the activity"
```
It should not be necessary to restart the activity.

quozl commented on frotz.py:
```diff
@@ -126,15 +131,45 @@ def start_game(self, game_file):
+                if response == Gtk.ResponseType.OK:
+                    logging.debug("Installing frotz")
+                    if platform.version().find("Ubuntu") > -1 or platform.version().find("Debian") > -1:
+                        cmd = "apt install frotz"
```
Doesn't the user have to be root in order to use these commands?  We don't run Sugar activities as root.

quozl commented on frotz.py:
```diff
@@ -143,19 +178,19 @@ def read_file(self, file_path):
-            os.chmod(tmpfolder, 0755)
-            os.chmod(tmpfilepath, 0755)
+            os.chmod(tmpfolder, 0o755)
+            os.chmod(tmpfilepath, 0o644)
```
https://github.com/sugarlabs/Frotz/pull/7/commits/607dd6ff8a5704fbe14a960443de12e6bddf9721 doesn't explain in the commit message _why_ you are changing the file mask.

Saumya-Mishra9129 replied:
https://github.com/sugarlabs/Frotz/pull/7/commits/69f178133e99c6ec2001a187df50a027eb8e997a Please see

quozl commented on frotz.py:
```diff
@@ -187,30 +222,54 @@ def _paste_cb(self, button):
+                None,
+                None)
+        else:
+            self.fork_command_full(
```
Why is it necessary to support versions of Vte without spawn_sync?

Saumya-Mishra9129 replied:
Actually I took terminal activity as reference. Terminal uses fork_command_full and spawn_sync https://github.com/sugarlabs/terminal-activity/blob/master/terminal.py#L496.

quozl reviewed (commented):
Thanks.  Reviewed again.

quozl commented:
Tested on Ubuntu 20.04;

* [ ] the stop button is in the wrong place,
* [ ] the dialog is not standard for Sugar, use an Alert class,
* [ ] clicking on "Get more games" just leads me to a Journal object for a link; it should launch Browse, (see Terminal for how),
* [ ] the more games link is external (at OLPC); it should be hosted in the repository,
* [ ] the install prepares a command but does not execute it,
* [ ] the install does not work, Permission denied,
* [ ] the font is very small, and is not the font configured by Sugar,
* [ ] there's no way to change the font size in the activity,

Saumya-Mishra9129 committed:
Use str as argument in ConfigParser

ConfigParser accepts <class 'str'> as *args
Boolean and integer values should be converted to str

Saumya-Mishra9129 committed:
Update README.md

Saumya-Mishra9129 committed:
Modify Quit method

Fix TypeError:_quit_cb takes 1-2 argument, 3 given
change sys.exit call to activity.close

Saumya-Mishra9129 committed:
Fix Indentation

Saumya-Mishra9129 commented:
> the more games link is external (at OLPC); it should be hosted in the repository,

What do you mean by it? I couldn't get completely , Is it mean including it in README.?

Saumya-Mishra9129 committed:
Add seperator before stop button

quozl commented:
Thanks.

I've tested 9dcd6b3.  Changing font size confuses the output.  I've reproduced without this activity; it happens when the virtual terminal changes size as consequence of font size changes.  I've [asked upstream](https://gitlab.com/DavidGriffith/frotz/-/issues/220).  For the moment, you can remove the font size changes.

Please use sugar3.graphics.style to set the FONT_SIZE.  Finance, Maze, Poll, and Words all use the Sugar setting.

> > the more games link is external (at OLPC); it should be hosted in the repository,
> 
> @quozl What do you mean by it? I couldn't get completely , Is it mean including it in README.?

At the moment the link is at wiki.laptop.org, which was set up in 2008 when the activity was first written.  In moving the activity to Sugar Labs all relevant sources and information should be in Sugar Labs infrastructure if possible.  GitHub holds the README.md file for us.

Regarding the Vte settings; parts of the activity were copied from Terminal as of 2008, probably so that settings changes made to Terminal would apply to Frotz as well.  I don't think all the settings are useful.  Remove as much as you can.

chimosky commented on frotz.py:
```diff
@@ -139,15 +139,17 @@ def start_game(self, game_file):
+                            save_dir,
+                            os.path.basename(game_file),
+                            game_file)
+
```
Why the empty line?

Saumya-Mishra9129 replied:
Removed in c7558d0

chimosky commented on frotz.py:
```diff
@@ -125,6 +134,37 @@ def __init__(self, handle):
+        return view_toolbar
+
+    def _zoom(self, step):
+
```
Why the empty line?

Saumya-Mishra9129 replied:
Thanks now I have removed this commit. Please See https://github.com/sugarlabs/Frotz/pull/7#issuecomment-675881114

Saumya-Mishra9129 committed:
Remove empty line caused in https://github.com/sugarlabs/Frotz/pull/7/commits/480c2079cdc483c7ea815bd0fd9e7cd23f752e81

Saumya-Mishra9129 committed:
Add font size configured by sugar

Saumya-Mishra9129 committed:
Use sugar.graphics.alert class to create an alert

Saumya-Mishra9129 committed:
Add requirement of sudo privilege while installing frotz

Saumya-Mishra9129 commented:
@quozl Please review till 1fe9fbb.

quozl commented:
Upstream suggested Fizmo.

quozl committed:
Review - README.md introduction

* fix typos,

* discriminate between Frotz and activity,

quozl committed:
Review - README.md dependencies

Introduction duplicated a list item.

quozl committed:
Review - README.md wrap

quozl committed:
Review - README.md add games

quozl commented:
Am reviewing.  Standby.

quozl committed:
Review - reapply license

Reapply the license to the program, see "How to Apply These Terms to
Your New Programs" in COPYING.

quozl committed:
Review - restore spacing

quozl committed:
Review - activity button is to be first

quozl commented:
* [ ] the start button is misleading; it does nothing except in the scenario where frotz is not installed; rather than use feed child to send install commands to the terminal, use a subprocess, or packagekit.

Saumya-Mishra9129 committed:
Add alert to start a game after installation of frotz

Saumya-Mishra9129 commented:
> the start button is misleading; it does nothing except in the scenario where frotz is not installed; rather than use feed child to send install commands to the terminal, use a subprocess, or packagekit.

Thanks @quozl It is fixed in 52f914c.

Saumya-Mishra9129 commented:
> clicking on "Get more games" just leads me to a Journal object for a link; it should launch Browse, (see Terminal for how),

Only this needs a fix.

Saumya-Mishra9129 commented:
a164e54 It does not work this error in log:

```Traceback (most recent call last):
  File "/usr/share/sugar/activities/frotz/frotz.py", line 247, in _get_games_cb
    subprocess.Popen(cmd, shell=False)
  File "/usr/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'xdg-open http://wiki.laptop.org/go/Frotz/Games': 'xdg-open http://wiki.laptop.org/go/Frotz/Games'```

quozl commented:
I suggest using launch_bundle as used by sugarterm.py in Terminal activity.

Saumya-Mishra9129 committed:
Fix: Click on button(Get more games) to launch Browse

Saumya-Mishra9129 commented:
> I suggest using launch_bundle as used by sugarterm.py in Terminal activity.

Thanks @quozl Done in ca59b68.:smile:

srevinsaju commented on frotz.py:
```diff
@@ -126,58 +138,97 @@ def start_game(self, game_file):
+            if platform.version().find("Ubuntu") > -1 or platform.version().find("Debian") > -1:
+                cmd = '/usr/games/frotz'
+            else:
+                cmd = '/usr/bin/frotz'
```
This still may not work. If I build frotz manually, it always installs to `/usr/local/bin` or `/usr/local/games` (on `apt` based distros). A suggested fix is to do this, as @quozl suggested is to:

https://github.com/sugarlabs-appstore/aslo-v4/blob/41383de6fa580dff1a6863c5d1505d01d5d25050/aslo4/platform/__init__.py#L29-L48
```python
def get_executable_path(executable, raise_error=True):
    """
    Returns the absolute path of the executable
    if it is found on the PATH,
    if it is not found raises a FileNotFoundError
    :param executable:
    :return:
    """

    for i in PATH:
        if os.path.exists(os.path.join(i, executable)):
            return os.path.join(i, executable)
    if raise_error:
        raise FileNotFoundError(
            "Could not find {p} on PATH. "
            "Make sure {p} is added to path and try again".format(
                p=executable))
    else:
        return False
```

Link to original source provided above the code snippet. 
You can do a `try...catch` on `FileNotFoundError` or pass `raise_error=False` and check if the returned value is false to check if Frotz exists on path.

Saumya-Mishra9129 replied:
Thanks , Tried it in 65c66e6. Tested on debian and Ubuntu works fine. :smiley:

srevinsaju replied:
Perfect! :tada:

Saumya-Mishra9129 committed:
Use PATH variable to find where frotz binary exists

quozl committed:
Review - superfluous import

quozl committed:
Review - encoding a constant

quozl committed:
Review - Frotz is not a module

quozl committed:
Review - use shutil.which

Python source was searched for references to the PATH environment
variable.  Of those available, the simplest API is shutil.which.

quozl committed:
Review - move common start code

quozl committed:
Review - point to repository for downloading games

An alternative would be to convert README.md to HTML and point the
browser to file:///$SOURCE_PATH/README.md.html

quozl committed:
Review - opening URL

- no need for another function,

- no need for "w+",

- change journal object description.
